docs/adrs/00007-purl-report.md

```diff
@@ -0,0 +1,104 @@
+re-use that functionality.
+
+As processing the correlation information may require some time, we may encounter request timeouts, like we sometimes
+see for larger SBOMs. The idea is to leverage the same pattern as we use for the UI upload. Upload with a `watch=true`
+flag, and then poll another API for the outcome.
+
+### Assumptions
+
+This ADR makes the following assumptions:
+
+* The provided SBOM will only be scanned for PURLs, not for CPEs, hashes/digests or other references
+
+## Decision
+
+There's a two-step process. First, upload (but don't store) the SBOM to the backend. The backend will parse the
+document and extract all PURLs, returning them to the client. This process, depending on the size of the file, is
+rather fast. The request shouldn't time out, as either the document or the response is being transferred. The actual
+processing, without any transfer, should be quick.
+
+The client can also list the result, once returned, to the user. To show the user what was understood from the SBOM.
+But that's an optional approach.
+
+Next, the client will open a websocket connection to the backend. It can then send as many PURLs as it wants to the 
+backend, which will look up vulnerabilities one by one and report them back to the client once correlated.
+
+## API
+
+* `POST /api/v2/extractPurlsFromSbom`: Send an SBOM, retrieve all PURLs
+
+  Response (`200 OK`):
+
+  ```json5
+  {
+    "purls": [
+      "pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources",
+      "pkg:npm/%40angular/animation@12.3.1"
+    ]
+  }
+  ```
+  
+* `POST /v2/vulnerability/analyze`: Analyze PURLs (already exists)
+
+  Request:
+
+  ```json5
+  {
+    "purls": [
+      "pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources",
+      "pkg:npm/%40angular/animation@12.3.1"
+    ]
+  }
+  ```
+
+  Response: (`200 OK`)
+
+  ```json5
+  {
+    "pkg:npm/%40angular/animation@12.3.1": [
+      {} // vulnerability information
+    ],
+    "pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources": [
+      {},  // vulnerability information
+      {}   // vulnerability information
+    ]
+  }
+  ```
+
+## Open items
+
+* Find a better path/name for the "extract PURLs" endpoint 
+
+## Alternative approaches
+
+Have the browser extract the PURLs and send those to the backend for checking.
+
+* 👍 The data transmitted to the backend would be way less. This would also need less CPU and memory on the backend.
+* 👍 The full data would never end up on the backend, so we'd never get in contact with information we don't need
+* It's unclear (without measuring) what's faster from a user perspective. Processing the file locally requires resources
+  and takes some time. But transmitting a larger document does too.
+* 👎 There's a notion of consolidating logic on the backend.
+
+Have the browser use a websocket API instead of the DA HTTP API:
+
+* 👎 Add "websockets" as a technology
+* 👎 Duplicate functionality of the dependency analytics API
```

Comment:
Agree.
However, this also means that if the IDE sends the SBOM to DA and DA sends it to TPA that might sound excessive for such a trivial task (again, depending on the type of extraction).
IMO I would not touch the current DA functionality.
Also don't forget about the current JSON and HTML reports generated in DA (in case you are considering adding something similar to the UI)

Reply:
For DA, it's ok keeping it that. This ADR is about having the browser generate a report and visualizing this. @carlosthe19916 needs to let us know what the UI requires for that from the backend.

I'd prefer the browser to extract that information too. As I think it removes unnecessary load from the backend. And, as you said, DA duplicates that already to some degree.

Which brings up the question: could we unify that between the browser and DA? Could we come up with a backend endpoint supporting both. Minimum the JSON. At best, also the HTML. :point_left: @carlosthe19916 @ruromero

Reply:
Added this to the open items section.

Reply:
In the Trustification UI use case, it doesn't make sense to generate the HTML report in the server. DA does that because there is no web client as such.
DA will not need to send the SBOM to trustification, it will only request the vulnerabilities by purl as it already does.

Reply:
I was just agreeing with you on the fact that for trustify UI the HTML should be generated in the client.
The DA backend generates a unified HTML because there's no UI to execute in the client. It goes directly to a file that is persisted in the file system to be displayed by the IDE.

I wanted to clarify these 2 differences between the DA and the UI. No big deal :)

Reply:
As I mentioned before, I don't think DA should move this functionality to TPA, it is an unnecessary overhead for a functionality that is already in place and is not expected to change.